torch/csrc/jit/interpreter.cpp

```diff
@@ -204,24 +448,35 @@ struct InterpreterStateImpl {
   void runOneStage(
     const std::vector<at::Tensor> & inputs,
     std::vector<at::Tensor> & outputs) {
-      //std::cout << "running stage: " << current_stage << " of " << function->stages.size() << "\n";
+      // std::cout << "running stage: " << current_stage << " of " << function->stages.size() << "\n";
       JIT_ASSERT(current_stage < function->stages.size());
       auto & stage = function->stages[current_stage++];
       JIT_ASSERT((int)inputs.size() == stage.inputs.size);
       for(int i = 0; i < stage.inputs.size; i++) {
         int reg = get(stage.inputs,i);
         if(reg >= 0) { // otherwise this input is dead, and we do not store it to avoid holding the reference
-          registers[reg] = inputs[i];
+          registers.takeOwnership(reg, toRetainableShare(inputs[i]));
         }
-        //std::cout << "registers[" << reg << "] = inputs[" << i << "](" << inputs[i].defined() << ")\n";
+        // std::cout << "registers[" << reg << "] = inputs[" << i << "](" << registers[reg] << ")\n";
       }
       for(auto & inst : stage.instructions) {
-        loadTensorsFromRegisters(inst.inputs, input_buffer);
+        auto & inputs = inst.inputs.values;
+        for(int i = 0; i < inputs.size; i++) {
+          int reg = get(inputs,i);
+          input_buffer.push_back(registers[reg]);
```

Comment:
So this effectively creates a temporary, because it only copies pointers to retainables, correct?

Reply:
Yes, the input/output buffers are just short term places where inputs/outputs get stored before moving into registers. The output list is unmanaged, but during the time it is unempty the only things that happen are appending to the output list and writting the outputs back into registers.